File: suho/withjava/src/baekjoon/test0928/B1890.java
package baekjoon.test0928;

import java.io.*;
import java.util.StringTokenizer;

public class B1890 {
    static int[][] pane;
    static int n;
    static long[][] answer;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;
        n = Integer.parseInt(br.readLine());
        pane = new int[n][n];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                pane[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        answer = new long[n][n];
        answer[0][0] = 1;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == n - 1 && j == n - 1) continue;
                int nextR = i + pane[i][j];
                int nextC = j + pane[i][j];

                if (0 <= nextR && nextR < n) {
                    answer[nextR][j] += answer[i][j];
                }

                if (0 <= nextC && nextC < n) {
                    answer[i][nextC] += answer[i][j];
                }
            }
        }
        bw.write(answer[n-1][n-1] + "");
        bw.flush();
        bw.close();
        br.close();
    }
}

File: jaeseuk/kokotlin/src/main/kotlin/boj/week5/B1890.kt
package boj.week5

//  점프

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val n = br.readLine().toInt()
    val board = Array(n) { StringTokenizer(br.readLine()).run { IntArray(n) { nextToken().toInt() } } }

    val dp = Array(n) { LongArray(n) }
    dp[0][0] = 1

    for (i in 0 until n) {
        for (j in 0 until n) {
            if (i == n - 1 && j == n - 1) break

            if (dp[i][j] > 0) {
                val right = j + board[i][j]
                val down = i + board[i][j]

                if (right < n) dp[i][right] += dp[i][j]
                if (down < n) dp[down][j] += dp[i][j]
            }
        }
    }

    bw.append("${dp[n - 1][n - 1]}")
    bw.flush()
    bw.close()
}